src/librustc/ty/sty.rs

```diff
@@ -979,29 +979,21 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
-    /// Checks whether a type is uninhabited.
-    /// If `block` is `Some(id)` it also checks that the uninhabited-ness is visible from `id`.
-    pub fn is_uninhabited(&self, block: Option<NodeId>, cx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
+    /// Checks whether a type is visibly uninhabited from a particular module.
+    pub fn is_uninhabited_from(&self, module: DefId, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
         let mut visited = FxHashSet::default();
-        self.is_uninhabited_recurse(&mut visited, block, cx)
+        let forrest = self.uninhabited_from(&mut visited, tcx);
+        forrest.contains(tcx, module)
     }
 
-    pub fn is_uninhabited_recurse(&self,
-                                  visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
-                                  block: Option<NodeId>,
-                                  cx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
-        match self.sty {
-            TyAdt(def, substs) => {
-                def.is_uninhabited_recurse(visited, block, cx, substs)
-            },
-
-            TyNever => true,
-            TyTuple(ref tys) => tys.iter().any(|ty| ty.is_uninhabited_recurse(visited, block, cx)),
-            TyArray(ty, len) => len > 0 && ty.is_uninhabited_recurse(visited, block, cx),
-            TyRef(_, ref tm) => tm.ty.is_uninhabited_recurse(visited, block, cx),
-
-            _ => false,
-        }
+    /// Checks whether a type is uninhabited.
+    /// Note: just because a type is uninhabited, that doesn't mean that it's
+    /// *visibly* uninhabited outside its module. You sometimes may want
+    /// `is_uninhabited_from` instead.
+    pub fn is_uninhabited_anywhere(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
```

Comment:
I can, but I'm pretty sure we *should* be using it. When I implemented `!` I left comments like this scattered throughout the code

    // FIXME(canndrew): This is_never should really be an is_uninhabited

The idea was that in most the places where we checked for the old `FnDiverges` or the new `TyNever` we should really be checking for uninhabited types generally. As far as I'm concerned `!` is just `enum Void` except it has its own syntax and it coerces. So I'm not sure why `!` still gets special treatment in a lot of places throughout the compiler. Example: greping for that comment takes me first to this code in liveness.rs

    hir::ExprCall(ref f, ref args) => {
      // FIXME(canndrew): This is_never should really be an is_uninhabited
      let diverges = !self.ir.tcx.tables().is_method_call(expr.id) &&
          self.ir.tcx.tables().expr_ty_adjusted(&f).fn_ret().0.is_never();
      let succ = if diverges {
          self.s.exit_ln
      } else {
          succ
      };
      let succ = self.propagate_through_exprs(args, succ);
      self.propagate_through_expr(&f, succ)
    }

So it looks like `!` and `Void` get treated differently by liveness even though they effect liveness identically.

---

For now I've removed the function but left a comment hinting at how it could be re-implemented.